mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp

```diff
@@ -0,0 +1,789 @@
+    if (lhs.isa<AffineDimExpr>()) {
+      auto dimExpr = lhs.cast<AffineDimExpr>();
+      lhsPosition = loopIndexMap[operands[dimExpr.getPosition()]];
+    }
+    // Update the loopIV only if it has not been encountered before. Please note
+    // that it is possible that the same loopIV have been encountered before
+    // while parsing other exprs. In that case, the appropriate coefficient is
+    // already set.
+    if (row[lhsPosition] == 0)
+      row[lhsPosition] = 1;
+    // The `rhs` may be a constant expr. In that case, no need to update the
+    // `row`.
+    bool isConstRhs = false;
+    if (rhs.isa<AffineDimExpr>()) {
+      auto dimExpr = rhs.cast<AffineDimExpr>();
+      rhsPosition = loopIndexMap[operands[dimExpr.getPosition()]];
+    } else if (rhs.isa<AffineConstantExpr>()) {
+      row.back() += rhs.cast<AffineConstantExpr>().getValue();
+      isConstRhs = true;
+    }
+    if (row[rhsPosition] == 0 && !isConstRhs)
+      row[rhsPosition] = 1;
+    break;
+  }
+  case AffineExprKind::Mul: {
+    AffineBinaryOpExpr mulExpr = expr.cast<AffineBinaryOpExpr>();
+    AffineExpr lhs = mulExpr.getLHS();
+    AffineExpr rhs = mulExpr.getRHS();
+    unsigned dimIdPos = 0;
+    // In the case of a mul expr, the lhs can only be a dim expr and the rhs can
+    // only be a constant expr.
+    if (lhs.isa<AffineDimExpr>()) {
+      auto dim = lhs.cast<AffineDimExpr>();
+      dimIdPos = loopIndexMap[operands[dim.getPosition()]];
+    }
+    if (rhs.isa<AffineConstantExpr>()) {
+      row[dimIdPos] = rhs.cast<AffineConstantExpr>().getValue();
+    }
+    break;
+  }
+  case AffineExprKind::DimId: {
+    // This takes care of the cases like A[i] where i is a loopIV. Since it is
+    // not a binary expr, there is no lhs/rhs.
+    auto dimExpr = expr.cast<AffineDimExpr>();
+    row[loopIndexMap[operands[dimExpr.getPosition()]]] = 1;
+    break;
+  }
+  case AffineExprKind::CeilDiv:
+  case AffineExprKind::FloorDiv:
+  case AffineExprKind::Mod: {
+    // Even though exprs like CeilDiv/FloorDiv and Mod can be considered as
+    // binary exprs, the `rhs` in these exprs is always a constant as per the
+    // rules of AffineExpr. These constant values do not make part of either
+    // the vector-b or the matrix A. Thus, we don't need to care about `rhs`
+    // in these cases.
+    auto modExpr = expr.cast<AffineBinaryOpExpr>();
+    AffineExpr lhs = modExpr.getLHS();
+    if (lhs.isa<AffineDimExpr>()) {
+      auto dimExpr = lhs.cast<AffineDimExpr>();
+      row[loopIndexMap[operands[dimExpr.getPosition()]]] = 1;
+    }
+  }
+  }
+}
+
+/// Populates `loopAccessMatrices` with the access matrices (A|b) of all load 
+/// and store ops in the loop body. Please note that each affine access can be 
+/// represented as a linear system Ax+b (A is the affine access matrix, x is the 
+/// vector of loopIVs and b is the constant-term vector). `loopIndexMap` holds 
+/// depth locations of each loopIV in the original loop order.
```

Comment:
This isn't true when you have floordiv/mod in accesses. What's the strategy in those cases. The test cases are also missing anything with a mod/div.